### ISSUE TITLE
octo.repos(OWNER, NAME) object is different than the result of user.repos.fetch()

### ISSUE THREAD

**Issue by miztroh** [philschatz/octokat.js#17]
How should repo.contents be used?  Any time I call repo.contents(), it always returns the files / folders from the root of the default branch.  How can I specify a different branch and / or path?

**Comment by miztroh** [comment]
I can get it to work this way.

```octo.repos('user', 'repo').contents('path?ref=branch').fetch().then(function () {console.log(arguments);});```

How can I get it to work this way?

```octo.me.fetch().then(function (user) {user.repos.fetch().then(function (repos) {repos[0].contents('path?ref=branch');```

**Comment by philschatz** [comment]
I believe the following should work:

```js
repo.contents('README.md').fetch({ref: 'my-branch'}) ...
```

or, alternatively:

```js
repo.contents.fetch({path: 'README.md', ref: 'my-branch'}) ...
```

(from <https://developer.github.com/v3/repos/contents/#get-contents>)

**Comment by miztroh** [comment]
In this example, repos[0].contents.fetch is undefined;

```octo.me.fetch().then(function (user) {user.repos.fetch().then(function (repos) {repos[0].contents.fetch;});```

**Comment by philschatz** [comment]
Yep, the `octo.repos(OWNER, NAME)` object is different than the result of `user.repos.fetch()`.

Can you try something like the following:

```js
octo.me.fetch().then(function (user) {
  user.repos.fetch().then(function (repos) {
    repo = repos[0];
    octo.repos(repo.owner.login, repo.name).contents(path).fetch({ref:'branch'}).then(function (contents) {
      ...
```

**Comment by miztroh** [comment]
Yeah, I just thought of that workaround.  Shouldn't you get the same type of repo object either way?

**Comment by miztroh** [comment]
Seems to me that the repo objects returned from either of these methods should be identical in functionality.

**Comment by philschatz** [comment]
or just:

```js
octo.me.repos.fetch().then(function (user) {
    repo = repos[0];
    octo.repos(repo.owner.login, repo.name).contents(PATH).fetch({ref:BRANCH_NAME}).then(function (contents) {
      ...
```

`octo.repos(...)` is constructed locally and the resulting `repos` object is constructed from the JSON (which looks similar to <https://developer.github.com/v3/repos/#get>) using URL Templates.

Every key ending in `_url` is converted to a fetch method so you can do `repos[0].owner.organizations()` but since there is no `content_url: "https://api.github.com/repos/philschatz/octokat.js/contents{/path}"` there is no `.contents()` on that object.

**Comment by philschatz** [comment]
There is an `OBJECT_MATCHER` that attempts to provide the same operations on a Repo JSON fragment (which is what I think you were expecting).

What kind of error did you get originally? Did you get a `BUG: Invalid Path.` Error, `undefined is not a function`, or some other error?

**Comment by miztroh** [comment]
Maybe you could special case this then.  I can use the workaround, but I think it should work either way.

Consider the logic of a web app using this library:
1. Login (user = result of octo.me)
2. List user repos (repos = result of user.repos)
3. Select a repo (repo = repos[SELECTED_REPO_INDEX])
4. List repo contents (repo.contents)

I'd submit a pull request for this myself, but I have no experience in CoffeeScript.

**Comment by philschatz** [comment]
Ah, now i see. The following works but without the branch option:

```js
octo.me.repos.fetch().then(function(repos) {
  repos[0].contents('README.md').then(function(x) {
    console.log(x);
  })
})
```

It's using the `contents()` from the Hypermedia URL. That could be a bug. 

Thanks! I'll try to take a closer look over the weekend.

**Comment by miztroh** [comment]
@philschatz Any update on this?

**Comment by philschatz** [comment]
Unfortunately, not yet. I spent last weekend getting the Travis-CI tests running again. I might have some time Monday to take a look.

**Comment by philschatz** [comment]
Thanks again; it should be fixed in `octokat@0.4.6`
